File: apps/enterprise-web/src/api/enterprise-support-api.ts
import type { EnterpriseSupportRagResponse } from "@translation/contracts";
import type { EnterpriseContentRequestContext } from "./enterprise-api.js";

type EnterpriseRequester = <T>(path: string, init?: RequestInit) => Promise<T>;

export interface EnterpriseSupportQueueDto {
  id: string;
  name: string;
  status: "active" | "paused" | "disabled";
  defaultPriority: number;
  handoffSlaSeconds: number;
  claimLeaseSeconds: number;
  version: number;
}

export interface EnterpriseSupportWorkItemDto {
  sessionId: string;
  queueId: string;
  customerId: string;
  priority: number;
  intent?: string;
  status: "handoff_requested" | "claim_expired";
  handoffRequestedAt: string;
  slaDeadlineAt: string;
  slaBreached: boolean;
  waitSeconds: number;
  expectedSessionVersion: number;
}

export interface EnterpriseSupportClaimDto {
  id: string;
  queueId: string;
  agentUserId: string;
  status: "active" | "released" | "reassigned" | "expired";
  claimedAt: string;
  leaseExpiresAt: string;
  updatedAt: string;
  version: number;
}

export interface EnterpriseSupportSessionDto {
  id: string;
  status: string;
  intent?: string;
  priority: number;
  queueId?: string;
  assignedUserId?: string;
  createdAt: string;
  queuedAt?: string;
  handoffRequestedAt?: string;
  assignedAt?: string;
  updatedAt: string;
  version: number;
}

export interface EnterpriseSupportFollowupDto {
  id: string; kind: "ticket" | "callback";
  status: "processing" | "completed" | "failed";
  caseId?: string; callbackId?: string; providerSimulated: boolean;
  attempts: number; failureCode?: string; createdAt: string;
  updatedAt: string; completedAt?: string; version?: number;
}

export interface EnterpriseSupportWorkbenchDto {
  generatedAt: string;
  session: EnterpriseSupportSessionDto;
  claim: EnterpriseSupportClaimDto;
  aiSpeechFence: {
    status: "stopped" | "not_started" | "terminal";
    verifiedAt: string;
    runId?: string;
    runStatus?: string;
  };
  channel: { channelType: string; provider: string; status: string };
  customer: {
    id: string; externalId?: string; displayName?: string; locale?: string;
    attributes: Record<string, unknown>; consentScope: string[]; updatedAt: string;
  };
  queue?: {
    id: string; name: string; status: string; handoffSlaSeconds: number;
    claimLeaseSeconds: number;
  };
  communication?: {
    sessionId: string; status: string; generation: number; updatedAt: string;
  };
  cases: Array<{
    id: string; subject: string; status: string; summary?: string;
    resolution?: string; externalTicketId?: string; createdAt: string;
    updatedAt: string;
  }>;
  callbacks: Array<{
    id: string; scheduledAt: string; reason: string;
    status: "dispatch_pending" | "scheduled" | "failed" | "cancelled" | "completed";
    externalCallbackId?: string; failureCode?: string; createdAt: string;
    updatedAt: string; completedAt?: string;
  }>;
  followups: EnterpriseSupportFollowupDto[];
  toolExecutions: Array<{
    id: string; toolName: string; riskLevel: string;
    confirmationStatus: string; status: string; resultDocument?: unknown;
    failureCode?: string; createdAt: string; completedAt?: string; updatedAt: string;
  }>;
  agent?: {
    runId: string; status: string; locale: string; countryCode: string;
    productCode: string; conversationState: string; generation: number;
    updatedAt: string;
  };
  conversationContext: Array<{ role: "customer" | "assistant"; text: string }>;
  agentTurns: Array<{
    id: string; sequence: number; status: string; spokenText: string;
    intent: string; riskSignals: string[]; knowledgeCitations: string[];
    createdAt: string; updatedAt: string;
  }>;
  highRiskHandoffs: Array<{
    id: string; toolName: string; riskCategory: string;
    policyVersion: string; createdAt: string;
  }>;
  transcriptSegments: Array<{
    segmentId: string; revision: number; sourceText: string;
    translatedText?: string; sourceLanguage?: string; targetLanguage?: string;
    speakerId?: string; speakerRole?: string; startMs?: number; endMs?: number;
    createdAt: string; updatedAt: string;
  }>;
  controls: Record<string, { status: "ready" | "forbidden" | "not_ready";
    reasonCode?: string; simulated?: boolean }>;
}

export interface EnterpriseSupportApi {
  listSupportQueues(context: EnterpriseContentRequestContext): Promise<{
    queues: EnterpriseSupportQueueDto[];
  }>;
  listSupportWorkItems(
    context: EnterpriseContentRequestContext,
    queueId: string,
  ): Promise<{ status: "ready"; workItems: EnterpriseSupportWorkItemDto[] }>;
  claimSupportSession(
    context: EnterpriseContentRequestContext,
    sessionId: string,
    input: { expectedSessionVersion: number; idempotencyKey: string },
  ): Promise<{ status: "claimed" | "replayed"; claim: EnterpriseSupportClaimDto;
    session: EnterpriseSupportSessionDto }>;
  activateSupportWorkbench(
    context: EnterpriseContentRequestContext,
    sessionId: string,
  ): Promise<EnterpriseSupportWorkbenchDto>;
  getSupportWorkbench(
    context: EnterpriseContentRequestContext,
    sessionId: string,
  ): Promise<EnterpriseSupportWorkbenchDto>;
  renewSupportClaim(
    context: EnterpriseContentRequestContext,
    claimId: string,
    expectedClaimVersion: number,
  ): Promise<{ claim: EnterpriseSupportClaimDto }>;
  releaseSupportClaim(
    context: EnterpriseContentRequestContext,
    claimId: string,
    input: { expectedClaimVersion: number; expectedSessionVersion: number;
      idempotencyKey: string; reason: "agent_release" | "agent_disconnect" },
  ): Promise<{ status: "released" | "replayed"; claim: EnterpriseSupportClaimDto;
    session: EnterpriseSupportSessionDto }>;
  createSupportTicket(
    context: EnterpriseContentRequestContext, sessionId: string,
    input: { subject: string; description: string; idempotencyKey: string;
      expectedSessionVersion: number; expectedClaimVersion: number },
  ): Promise<{ status: "processing" | "replayed";
    followup: EnterpriseSupportFollowupDto }>;
  scheduleSupportCallback(
    context: EnterpriseContentRequestContext, sessionId: string,
    input: { scheduledAt: string; reason: string; idempotencyKey: string;
      expectedSessionVersion: number; expectedClaimVersion: number },
  ): Promise<{ status: "processing" | "replayed";
    followup: EnterpriseSupportFollowupDto }>;
  resolveSupportKnowledge(
    context: EnterpriseContentRequestContext,
    sessionId: string,
    input: { query: string; locale: string; countryCode: string;
      productCode: string; limit?: number },
  ): Promise<{ resolution: EnterpriseSupportRagResponse }>;
}

export function createEnterpriseSupportApi(
  request: EnterpriseRequester,
  headers: (context: EnterpriseContentRequestContext) => Record<string, string>,
): EnterpriseSupportApi {
  const sessionPath = (sessionId: string) =>
    `/enterprise/v1/support/sessions/${encodeURIComponent(sessionId)}`;
  return {
    listSupportQueues: (context) => request("/enterprise/v1/support/queues", {
      headers: headers(context),
    }),
    listSupportWorkItems: (context, queueId) => request(
      `/enterprise/v1/support/queues/${encodeURIComponent(queueId)}/work-items`,
      { headers: headers(context) },
    ),
    claimSupportSession: (context, sessionId, input) => request(
      `${sessionPath(sessionId)}/claims`,
      { method: "POST", headers: headers(context), body: JSON.stringify(input) },
    ),
    activateSupportWorkbench: (context, sessionId) => request(
      `${sessionPath(sessionId)}/workbench`,
      { method: "POST", headers: headers(context), body: JSON.stringify({}) },
    ),
    getSupportWorkbench: (context, sessionId) => request(
      `${sessionPath(sessionId)}/workbench`, { headers: headers(context) },
    ),
    renewSupportClaim: (context, claimId, expectedClaimVersion) => request(
      `/enterprise/v1/support/claims/${encodeURIComponent(claimId)}/renew`,
      { method: "POST", headers: headers(context),
        body: JSON.stringify({ expectedClaimVersion }) },
    ),
    releaseSupportClaim: (context, claimId, input) => request(
      `/enterprise/v1/support/claims/${encodeURIComponent(claimId)}/release`,
      { method: "POST", headers: headers(context), body: JSON.stringify(input) },
    ),
    createSupportTicket: (context, sessionId, input) => request(
      `${sessionPath(sessionId)}/followups/tickets`,
      { method: "POST", headers: headers(context), body: JSON.stringify(input) },
    ),
    scheduleSupportCallback: (context, sessionId, input) => request(
      `${sessionPath(sessionId)}/followups/callbacks`,
      { method: "POST", headers: headers(context), body: JSON.stringify(input) },
    ),
    resolveSupportKnowledge: (context, sessionId, input) => request(
      `${sessionPath(sessionId)}/rag`,
      { method: "POST", headers: headers(context), body: JSON.stringify(input) },
    ),
  };
}

File: apps/mobile/lib/src/features/enterprise/data/enterprise_support_api.dart
import 'enterprise_mobile_api_client.dart';
import 'enterprise_mobile_models.dart';
import 'enterprise_support_models.dart';
import 'enterprise_support_workbench_models.dart';

extension EnterpriseMobileSupportApi on EnterpriseMobileApiClient {
  Future<List<EnterpriseMobileSupportQueue>> listSupportQueues(
    EnterpriseMobileWorkspace workspace,
  ) async {
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/queues',
      method: 'GET',
    );
    final values = json['queues'];
    if (values is! List<Object?>) {
      throw const FormatException('Invalid enterprise support queues');
    }
    return values.map((item) {
      if (item is! Map<String, Object?>) {
        throw const FormatException('Invalid enterprise support queue');
      }
      return EnterpriseMobileSupportQueue.fromJson(item);
    }).toList(growable: false);
  }

  Future<List<EnterpriseMobileSupportWorkItem>> listSupportWorkItems(
    EnterpriseMobileWorkspace workspace,
    String queueId,
  ) async {
    if (!supportUuidValue(queueId)) {
      throw const FormatException('Invalid enterprise support queue id');
    }
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/queues/${Uri.encodeComponent(queueId)}/work-items',
      method: 'GET',
    );
    if (json['status'] != 'ready' || json['workItems'] is! List<Object?>) {
      throw const FormatException('Invalid enterprise support work items');
    }
    return (json['workItems']! as List<Object?>).map((item) {
      if (item is! Map<String, Object?>) {
        throw const FormatException('Invalid enterprise support work item');
      }
      final workItem = EnterpriseMobileSupportWorkItem.fromJson(item);
      if (workItem.queueId != queueId) {
        throw const EnterpriseMobileApiException(
          code: 'tenant_context_mismatch',
          message: 'Enterprise support queue mismatch',
        );
      }
      return workItem;
    }).toList(growable: false);
  }

  Future<EnterpriseMobileSupportClaimResult> claimSupportSession(
    EnterpriseMobileWorkspace workspace,
    EnterpriseMobileSupportWorkItem workItem,
    String idempotencyKey,
  ) async {
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/sessions/'
      '${Uri.encodeComponent(workItem.sessionId)}/claims',
      method: 'POST',
      body: <String, Object?>{
        'expectedSessionVersion': workItem.expectedSessionVersion,
        'idempotencyKey': idempotencyKey,
      },
    );
    final result = EnterpriseMobileSupportClaimResult.fromJson(json);
    _validateClaim(
      workspace,
      result.claim,
      result.session,
      sessionId: workItem.sessionId,
      queueId: workItem.queueId,
    );
    return result;
  }

  Future<EnterpriseMobileSupportWorkbench> activateSupportWorkbench(
    EnterpriseMobileWorkspace workspace,
    String sessionId,
  ) async {
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/sessions/${Uri.encodeComponent(sessionId)}/workbench',
      method: 'POST',
      body: const <String, Object?>{},
    );
    return _workbench(workspace, sessionId, json);
  }

  Future<EnterpriseMobileSupportWorkbench> getSupportWorkbench(
    EnterpriseMobileWorkspace workspace,
    String sessionId,
  ) async {
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/sessions/${Uri.encodeComponent(sessionId)}/workbench',
      method: 'GET',
    );
    return _workbench(workspace, sessionId, json);
  }

  Future<EnterpriseMobileSupportClaim> renewSupportClaim(
    EnterpriseMobileWorkspace workspace,
    EnterpriseMobileSupportClaim claim,
  ) async {
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/claims/${Uri.encodeComponent(claim.id)}/renew',
      method: 'POST',
      body: <String, Object?>{'expectedClaimVersion': claim.version},
    );
    final renewed = EnterpriseMobileSupportClaim.fromJson(
      supportMap(json, 'claim'),
    );
    if (renewed.id != claim.id ||
        renewed.supportSessionId != claim.supportSessionId ||
        renewed.agentUserId != workspace.context.member.userId ||
        renewed.status != 'active') {
      throw const EnterpriseMobileApiException(
        code: 'tenant_context_mismatch',
        message: 'Enterprise support claim renewal mismatch',
      );
    }
    return renewed;
  }

  Future<EnterpriseMobileSupportReleaseResult> releaseSupportClaim(
    EnterpriseMobileWorkspace workspace,
    EnterpriseMobileSupportClaim claim,
    EnterpriseMobileSupportSession session,
    String idempotencyKey, {
    required String reason,
  }) async {
    if (!const <String>{'agent_release', 'agent_disconnect'}.contains(reason)) {
      throw const FormatException('Invalid enterprise support release reason');
    }
    final json = await contentRequest(
      workspace,
      '/enterprise/v1/support/claims/${Uri.encodeComponent(claim.id)}/release',
      method: 'POST',
      body: <String, Object?>{
        'expectedClaimVersion': claim.version,
        'expectedSessionVersion': session.version,
        'idempotencyKey': idempotencyKey,
        'reason': reason,
      },
    );
    final result = EnterpriseMobileSupportReleaseResult.fromJson(json);
    if (result.claim.id != claim.id ||
        result.claim.supportSessionId != session.id ||
        result.claim.agentUserId != workspace.context.member.userId ||
        result.claim.status != 'released' ||
        result.session.id != session.id) {
      throw const EnterpriseMobileApiException(
        code: 'tenant_context_mismatch',
        message: 'Enterprise support release mismatch',
      );
    }
    return result;
  }

  EnterpriseMobileSupportWorkbench _workbench(
    EnterpriseMobileWorkspace workspace,
    String sessionId,
    Map<String, Object?> json,
  ) {
    final workbench = EnterpriseMobileSupportWorkbench.fromJson(json);
    _validateClaim(
      workspace,
      workbench.claim,
      workbench.session,
      sessionId: sessionId,
      queueId: workbench.claim.queueId,
    );
    if (!workbench.aiSpeechFence.stopsNewAiSpeech) {
      throw const EnterpriseMobileApiException(
        code: 'ai_speech_fence_not_ready',
        message: 'Enterprise AI speech fence is not ready',
      );
    }
    return workbench;
  }

  void _validateClaim(
    EnterpriseMobileWorkspace workspace,
    EnterpriseMobileSupportClaim claim,
    EnterpriseMobileSupportSession session, {
    required String sessionId,
    required String queueId,
  }) {
    if (claim.supportSessionId != sessionId ||
        claim.queueId != queueId ||
        claim.agentUserId != workspace.context.member.userId ||
        session.id != sessionId ||
        session.queueId != queueId ||
        session.assignedUserId != workspace.context.member.userId ||
        (session.activeAgentClaimId != null &&
            session.activeAgentClaimId != claim.id) ||
        session.status != 'human_active' ||
        claim.status != 'active') {
      throw const EnterpriseMobileApiException(
        code: 'tenant_context_mismatch',
        message: 'Enterprise support claim context mismatch',
      );
    }
  }
}
